Urlaubsmodus into Develop

tobireith opened the pull request:
Der Urlaubsmodus hat nun volle Funktionalität:

Man kann Urlaube erzeugen (mit Name, Start- und Enddatum)
Daraus werden dann die Dauer in Tagen und die dadurch resultierende Ersparnis an Stromkosten berechnet und live angezeigt.
Beim anlegen eines Urlaubs wird immer von der aktuellen Gerätekonfiguration ausgegangen und diese kann später nicht mehr geändert werden.
Man kann Urlaube editieren, wobei man die Gerätekonfiguration nicht mehr ändern kann.

In der Übersicht wird der Name des Urlaubs, Startdatum und Enddatum angezeigt.

Bei der Auswertung werden Urlaube in dem Waterfall-Diagramm ebenso wie die Produzenten von dem Verbrauch abgezogen.

Außerdem wird der Gesamtverbrauch nun auch in Euro angezeigt (bisher nur in kWh). Der Jahresverbrauch pro Gerät wird weiterhin nur in kWh angezeigt (wenn man zusätzlich € anzeigen will, wäre wahrscheinlich ein anderes Diagramm oder ein weiteres Kuchendiagramm notwendig, da die Dateneinträge für ein Kuchendiagramm anscheinend nur einen Wert enthalten können).

tobireith committed:
Ersetzen des Home Fragments durch Urlaub Fragment mit Icon unt Text

tobireith committed:
Navigation zu neuen Urlaub anlegen und wieder zurück wurde hinzugefügt.
Die Navigation zu neuen Urlaub anlegen erfolgt, wie gewohnt, über den eingebundenen Floating Action button.
Zurück kommt man durch den Abbrechen Button

hoffmatteo committed:
Added urlaub entity

hoffmatteo committed:
Added urlaub entity v2

Commit:
Design: Neuen Urlaub anlegen, Refactor für einheitliche content_margin_top

Commit:
Funktion von Neuen Urlaub anlegen

Man kann nun einen neuen Urlaub anlegen, der dann auch in der DB gespeichert wird.
Bei änderung der Eingabe für das Datum wird automatisch die Dauer des Urlaubs und die Ersparnis in kWh und in € angezeigt.
Außerdem gibt es einen Hinweis, dass der Urlaub immer von der aktuellen Gerätekonfiguration ausgeht und diese später nicht mehr geändert werden kann. Der Hinweis beinhaltet auch Information darüber, wie die Ersparnis berechnet wird.

tobireith committed:
Urlaube werden nun in einer RecyclerView durch Name und Datum von & bis angezeigt.

tobireith committed:
Urlaube können nun bearbeitet werden, wobei sich die Konfiguration der Geräte nicht mehr ändert.
Mann kann Name, Start und Ende konfigurieren, aber die Konfiguration der Geräte bleibt die selbe wie zum Zeitpunkt des erstellens.

Einige Werte und Funktionen wurden in das UrlaubsCompainion ausgelagert, da diese beim bearbeiten und neu erstellen benötigt werden. (hier könnte man ggf. noch mehr auslagern)

Bestehende Urlaube kann man nun auch löschen

tobireith committed:
Urlaube werden nun nach Startdatum sortiert, FAB blockiert nicht mehr die untersten Urlaube

Außerdem wurde der Code Reformatted und einzelne Warnings behoben.

tobireith committed:
Fix: Das Programm stürzt nicht mehr ab, wenn kein Datum als String eingegeben wird
Der ganze Block wurde mit try, catch umschlossen um den Fehler gezielt zu behandeln.
Das Log ist beabsichtigt und wird anstatt dem StackTrace ausgegeben

tobireith committed:
Kleine Warning Fixes im UrlaubCompanion.kt

tobireith committed:
removed TODO

Commit:
Merge branch 'develop' into urlaubsmodus

# Conflicts:
#	app/schemas/com.example.stromtracker.database.AppDatabase/1.json
#	app/src/main/java/com/example/stromtracker/MainActivity.kt
#	app/src/main/java/com/example/stromtracker/database/AppDatabase.kt
#	app/src/main/java/com/example/stromtracker/database/DataRepository.kt
#	app/src/main/java/com/example/stromtracker/database/Geraete.kt
#	app/src/main/java/com/example/stromtracker/database/GeraeteDAO.kt
#	app/src/main/res/values/dimens.xml
#	app/src/main/res/values/strings.xml

Commit:
Kleine Fixes, Log entfernt bei Urlaubsmodus

Commit:
FIX: In der DB wird nun die Differenz des Tagesverbrauchs gespeichert, anstatt des neuen Tagesverbrauchs.
Ansonsten wären Rückschlüsse auf die Ersparnis nur möglich, wenn auch der bisherige Tagesverbrauch ohne Urlaub gespeichert werden würde.

Commit:
Rename in Database, changed name from "gesamtVerbrauch" to "ersparnisProTag", da jetzt dieser Wert gespeichert wird.

Commit:
Urlaube werden nun bei der Auswertung im Bilanzdiagramm (Waterfall-Chart) angezeigt
Dabei werden nur die Urlaube aus dem aktuellen Jahr miteinberechnet.

Commit:
Kommentar zur Bilanz hinzugefügt, Rename zu Bilanz anstatt Verbrauch-Produktion

Commit:
Fix für prozentualen Verlgeich mit Durchschnitt

Modulo 1 war nicht richtig, es ist -1 da sonst z.B. 166% als 66% angezeigt werden

tobireith committed:
Umstieg auf Livedata für Current Haushalt,

tobireith committed:
Umstieg auf LiveData für AuswertungsFragment, Deprecated ViewModelProviders Fix, Code Reformat, Newlines

emuguy1 commented on app/src/main/java/com/example/stromtracker/ui/geraete/auswertung/GeraeteAuswertungFragment.kt:
```diff
@@ -15,44 +14,44 @@ import com.anychart.AnyChartView
-import com.example.stromtracker.database.Haushalt
-import com.example.stromtracker.database.Kategorie
-import com.example.stromtracker.database.Raum
+import com.example.stromtracker.database.*
```
Sehr sinvoll. Sollten wir bei den anderen auch nochmal schauen, ob wir das nicht da auch machen wollen

tobireith replied:
Danke :)

emuguy1 commented on app/src/main/java/com/example/stromtracker/ui/geraete/auswertung/GeraeteAuswertungFragment.kt:
```diff
@@ -171,16 +166,29 @@ class GeraeteAuswertungFragment(
+        tempSum = tempList.sumByDouble { urlaub ->
+            urlaub.getErsparnisProTag() * (urlaub.getDateBis().time / UrlaubCompanion.dateTimeToDays - urlaub.getDateVon().time / UrlaubCompanion.dateTimeToDays + 1)
+        }
+        tempSum = tempSum.withSign(-1)
```
änderst du damit das vorzeichen des Wertes?

tobireith replied:
Das setzt das Vorzeichen auf Minus, egal welches es davor hatte

emuguy1 commented on app/src/main/java/com/example/stromtracker/ui/urlaub/UrlaubFragment.kt:
```diff
@@ -0,0 +1,108 @@
+        //RecyclerView initialisieren
+        viewAdapter = UrlaubListAdapter(urlaubList, sharedViewModel.getHaushalt().value!!)
+
+        viewManager = LinearLayoutManager(this.context)
+        recyclerView = root.findViewById<RecyclerView>(R.id.urlaub_recycler_view).apply {
+            setHasFixedSize(true)
+            layoutManager = viewManager
+            adapter = viewAdapter
+        }
```
Warum genau machst du das?

hoffmatteo replied:
Um die recyclerview zu instanziieren?

Habe ich bei Geräte genau so gemacht, machst du es anders oder wie?

emuguy1 replied:
Ich hab da nichts mit set setfixed size

hoffmatteo replied:
Hm, ist anscheinend für Optimierungen da.
https://stackoverflow.com/questions/28709220/understanding-recyclerview-sethasfixedsize
https://stackoverflow.com/questions/28827597/when-do-we-use-the-recyclerview-sethasfixedsize/40707099

Soweit ich verstehe, soll das true sein wenn sich die Größe der Recyclerview (nicht der Items) nicht aufgrund des Contents ändert.

Hast du bei dir die reyclerview auf wrap_content?

hoffmatteo replied:
Ah ne du hast es auf match_parent, interessant

hoffmatteo replied:
Ich glaube tatsächlich bei match_parent sollte es auf true sein, weil da die Größe bereits fest ist, maximal so groß wie der parent.

Bei wrap_content dagegen auf false, da es sich dann vergrößert je mehr Items dazukommen.

Schaut aber bitte ihr nochmal nach, nicht dass ich das falsch verstehe

hoffmatteo replied:
Wobei er das bei seinem Blatt nicht setzt, (und default ist false) obwohl er match_parent macht

emuguy1 commented on app/src/main/java/com/example/stromtracker/ui/urlaub/urlaub_edit/UrlaubEditFragment.kt:
```diff
@@ -0,0 +1,192 @@
+    }
+
+
+    private fun addCustomTextChangedListener(edit: EditText): EditText {
+        edit.addTextChangedListener(object : TextWatcher {
+            override fun afterTextChanged(s: Editable) {
+                if (checkDates(start, ende)) {
+                    val start =
+                        SimpleDateFormat(
+                            "dd.MM.yyyy",
+                            Locale.GERMAN
+                        ).parse(start.text.toString())
+                    val ende =
+                        SimpleDateFormat(
+                            "dd.MM.yyyy",
+                            Locale.GERMAN
+                        ).parse(ende.text.toString())
+                    updateOutputs(start, ende)
+                }
+            }
+
+            override fun beforeTextChanged(
+                s: CharSequence,
+                start: Int,
+                count: Int,
+                after: Int
+            ) {
+            }
+
+            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
+        })
+        return edit
+    }
```
Warum machst du hier einen TextListener

tobireith replied:
Der TextChangedListener ist dafür da um die Anzeige für die Dauer in Tagen und die daraus resultierende Ersparnis live anzuzuzeigen. 
Er triggert jedes mal, wenn eines der Datums geändert wird

emuguy1 replied:
Ok gut verstehe

emuguy1 commented on app/src/main/res/layout/fragment_geraete.xml:
```diff
@@ -312,4 +312,4 @@
 
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+</androidx.constraintlayout.widget.ConstraintLayout>
```
No newline?

tobireith replied:
Wird beim nächsten commit gefixt.

emuguy1 commented on app/src/main/res/layout/fragment_urlaub_edit.xml:
```diff
@@ -0,0 +1,161 @@
+        android:id="@+id/urlaub_edit_button_delete"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:layout_marginLeft="16dp"
+        android:layout_marginEnd="16dp"
+        android:layout_marginRight="16dp"
+        android:layout_marginBottom="16dp"
```
Hier hättest du noch einiges in dimens auslagern können. Und da du die API auf 19 gesetzt hast, kannst du das margin Left und Right weg lassen

tobireith replied:
Wird beim nächsten commit gefixt.

emuguy1 commented on app/src/main/res/layout/fragment_urlaub_new.xml:
```diff
@@ -0,0 +1,145 @@
+        android:id="@+id/urlaub_new_button_speichern"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="16dp"
+        android:layout_marginRight="16dp"
+        android:layout_marginBottom="16dp"
+        android:background="@color/colorPrimary"
+        android:text="@string/button_speichern"
+        android:textColor="@color/colorWhite"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent" />
+
```
Nochmal das gleiche

tobireith replied:
Wird beim nächsten commit gefixt.

emuguy1 commented on app/src/main/res/layout/fragment_urlaub_recycler_text_view.xml:
```diff
@@ -0,0 +1,75 @@
+        android:id="@+id/urlaub_recycler_card"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="8dp"
+        android:layout_marginLeft="8dp"
+        android:layout_marginTop="8dp"
+        android:layout_marginEnd="8dp"
+        android:layout_marginRight="8dp"
+        android:layout_marginBottom="8dp"
```
siehe oben

tobireith replied:
Wird beim nächsten commit gefixt.

emuguy1 commented on app/src/main/res/layout/fragment_urlaub_recycler_text_view.xml:
```diff
@@ -0,0 +1,75 @@
+    </androidx.cardview.widget.CardView>
+
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
no newline

tobireith replied:
Wird beim nächsten commit gefixt.

emuguy1 commented on app/src/main/res/navigation/mobile_navigation.xml:
```diff
@@ -54,9 +54,9 @@
+        android:label="@string/menu_urlaub"
+        tools:layout="@layout/fragment_urlaub" />
 
 </navigation>
```
no newline

tobireith replied:
Wird beim nächsten commit gefixt.

emuguy1 commented on app/src/main/res/menu/activity_main_drawer.xml:
```diff
@@ -49,8 +49,8 @@
+            android:icon="@drawable/ic_menu_urlaub"
+            android:title="@string/menu_urlaub" />
     </group>
 </menu>
```
no newline

tobireith replied:
Wird beim nächsten commit gefixt.

hoffmatteo commented on app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt:
```diff
@@ -296,7 +313,9 @@ class GeraeteFragment : Fragment(), View.OnClickListener {
-                    raumListHaushalt
+                    raumListHaushalt,
+                    urlaubList,
+                    sharedViewModel.getHaushalt().value!!
```
Warum greifst du hier ohne einen Observer auf die LiveData zu?

Den current Haushalt kannst du doch z.b aus der raumListHaushalt nehmen, da haben wir ja immer den einen Raum der auch die HaushaltID hat.

tobireith replied:
Habe es auf Observer umgeändert.

hoffmatteo commented on app/src/main/java/com/example/stromtracker/ui/geraete/auswertung/GeraeteAuswertungFragment.kt:
```diff
@@ -129,12 +124,12 @@ class GeraeteAuswertungFragment(
             pie.data(data)
-            gesamtverbrauch =
+            gesamtverbrauchkWh =
                 roundTo2Decimal(verbraucherList.sumByDouble { geraete -> geraete.getJahresverbrauch() })
```
Ich implementiere gerade, dass das nur gerundet in die Datenbank kommt, also sobald alles gemerged ist kann das dann weg.

tobireith replied:
Wird beim nächsten commit entfernt.

hoffmatteo commented on app/src/main/java/com/example/stromtracker/ui/urlaub/urlaub_edit/UrlaubEditFragment.kt:
```diff
@@ -0,0 +1,192 @@
+import com.example.stromtracker.R
+import com.example.stromtracker.database.Haushalt
+import com.example.stromtracker.database.Urlaub
+import com.example.stromtracker.ui.urlaub.UrlaubCompanion.Companion.centToEuro
+import com.example.stromtracker.ui.urlaub.UrlaubFragment
+import com.example.stromtracker.ui.urlaub.UrlaubCompanion.Companion.checkDates
+import com.example.stromtracker.ui.urlaub.UrlaubCompanion.Companion.dateTimeToDays
```
Das mit dem companion object ist gut, das werde ich auch so implementieren.

Muss man da aber jede Methode einzeln importen, oder geht das nicht mit Companion.*?

tobireith replied:
Ich musste jedes einzeln importieren, weil man das anscheinend nicht allgemein importieren kann.
Kann aber auch sein dass es da noch etwas schöneres gibt, mit import ...UrlaubCompanion.* hat es bei mir nicht geklappt

hoffmatteo commented on app/src/main/java/com/example/stromtracker/ui/urlaub/UrlaubFragment.kt:
```diff
@@ -0,0 +1,108 @@
+    override fun onClick(v: View) {
+        when (v.id) {
+            R.id.urlaub_button_add -> {
+                val frag = UrlaubNewFragment(verbraucherList, sharedViewModel.getHaushalt().value!!)
```
Siehe vorhin mit dem direkten Zugriff auf die LiveData

tobireith replied:
Habe es auf Observer umgeändert.

tobireith committed:
Änderungsvorschläge wurden übernommen
Einige Newlines hinzugefügt
Auslagerung in dimens.xml
Refactor von content_margin_sides
Zugriff auf LiveData nun über Observer

tobireith committed:
Merge branch 'develop' into urlaubsmodus

# Conflicts:
#	app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt
#	app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteViewModel.kt

hoffmatteo commented on app/src/main/java/com/example/stromtracker/ui/geraete/GeraeteFragment.kt:
```diff
@@ -156,6 +156,16 @@ class GeraeteFragment : Fragment(), View.OnClickListener {
                 }
             })
 
+        val currHaushaltData = Transformations.switchMap(sharedViewModel.getHaushalt()) { haushalt ->
+            geraeteViewModel.getHaushaltByID(haushalt.getHaushaltID())
+        }
+        currHaushaltData.observe(
+            viewLifecycleOwner,
+            Observer { haushalt ->
+                currHaushalt = haushalt
+            }
+        )
+
```
Warum machst du das über switchmap und nicht direkt einen Observer auf sharedViewModel.getHaushalt()?

Abgesehen davon brauchen wir ja diesen Observer nicht notwendigerweise, wir können die HaushaltID doch auch vom Standardraum ablesen, dann würden wir uns einen Observer sparen.

tobireith replied:
Ich benötige ja nicht die ID vom Haushalt sondern den Haushalt selbst, also einen Select-Befehl (Wir haben ja keine Haushaltsliste).
Okay, dann mache ich das mit der switchMap noch raus, dachte das braucht man irgendwie für das shared view Model.

tobireith committed:
Removed SwitchMap
